### PR TITLE
Move `SUBSCRIBE` optimization off the coordinator thread

### PR DESCRIPTION
**Pull request by aalexandrov**
Part of #16217 (this is part 5/5 for the "off-thread refactoring" for optimized statements).

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

This is similarly structured to #24049, #24068, #24072, and #24072.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/coord/sequencer/inner/subscribe.rs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+        let mut timeline = self.validate_timeline_context(depends_on.clone())?;
+        if matches!(timeline, TimelineContext::TimestampIndependent) && from.contains_temporal() {
+            // If the from IDs are timestamp independent but the query contains temporal functions
+            // then the timeline context needs to be upgraded to timestamp dependent.
+            timeline = TimelineContext::TimestampDependent;
+        }
+
+        // TODO: this check is present in `sequence_create_index` / `sequence_create_materialized_view`
+        // Should we have it here as well?
+        // self.ensure_cluster_can_host_compute_item(name, *cluster_id)?;
```
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner/subscribe.rs:150]: @jkosh44 I think you reviewed the PR that added this (#21846). LMK if it makes sense to add this check for `SUBSCRIBE` statements as well.

**Comment by aalexandrov** [comment]
@MaterializeInc/adapter whoever reviews this, if [the nightlies run](https://buildkite.com/materialize/nightlies/builds/5719) passes and the review lands before Jan 5th, feel free to amend the commits directly and merge the PR.

File: src/adapter/src/coord/sequencer/inner/subscribe.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/coord/sequencer/inner/subscribe.rs:150]: Good call out, but is this somehow related to moving optimization off of the main Coord thread? If not I'd prefer to move it to a separate PR.

**Comment by jkosh44** [comment]
Something that I think would be useful for this PR and all the others is a justification for how the existing `PlanValidity` is correct for all the new plans we've added over the past week. Most of our errors have been around `PlanValidity` not checking all the necessary metadata in `check()`.

Also do we have existing parallel tests that exercise all of the new plans that have been pulled off-thread?

**Review by jkosh44 — APPROVED** [review]

**Pull request merged**